### PR TITLE
move scopeCode to statementSem

### PR DESCRIPTION
**Pull request by benjones**
to get some semantic stuff out of the AST module as discussed in #12796

### CODE REVIEW

File: src/dmd/statementsem.d [hunk 1]
```diff
@@ -4473,6 +4473,95 @@ Statement semanticScope(Statement s, Scope* sc, Statement sbreak, Statement scon
...
+ * Returns:
+ *    code to be run in the finally clause
+ */
+Statement scopeCode(Statement statement, Scope* sc, Statement* sentry, Statement* sexception, Statement* sfinally)
+{
+    *sentry = null;
+    *sexception = null;
+    *sfinally = null;
+
+    if (auto es = statement.isExpStatement)
```
Comment (MoonlightSentinel) [inline src/dmd/statementsem.d:4495]:
```suggestion
    if (auto es = statement.isExpStatement())
```
Ditto for the other cases

File: src/dmd/statementsem.d [hunk 2]
```diff
@@ -4473,6 +4473,95 @@ Statement semanticScope(Statement s, Scope* sc, Statement sbreak, Statement scon
...
+        if (ls.statement)
+            ls.statement = ls.statement.scopeCode(sc, sentry, sexception, sfinally);
+        return ls;
+    }
+
+    return statement;
+}
+
+
+
```
Comment (MoonlightSentinel) [inline src/dmd/statementsem.d:4564]:
```suggestion

```

**Review by MoonlightSentinel — COMMENTED** [review]
Nice

File: src/dmd/statementsem.d [hunk 3]
```diff
@@ -4473,6 +4473,94 @@ Statement semanticScope(Statement s, Scope* sc, Statement sbreak, Statement scon
...
+ *     sexception = set to code executed upon exit from the scope via exception
+ *     sfinally   = set to code executed in finally block
+ * Returns:
+ *    code to be run in the finally clause
+ */
+Statement scopeCode(Statement statement, Scope* sc, Statement* sentry, Statement* sexception, Statement* sfinally)
+{
+    *sentry = null;
+    *sexception = null;
+    *sfinally = null;
```
Comment (ibuclaw) [inline src/dmd/statementsem.d:4493]: You could turn these into `out` parameters.

Comment (ibuclaw) [reply src/dmd/statementsem.d:4493]: Preferably in a follow-up, however. As that would alter the callers too.

Comment (benjones) [reply src/dmd/statementsem.d:4493]: Good call.  Can I do it in a new PR since I think will change a bunch of callsites?

**Comment by benjones** [comment]
Looks like the buildkite failure is related to dlangbot + sqlite3 which I think is a known issue?

**Comment by thewilsonator** [comment]
Yes, _that_ has been fixed, although I think there is still some problems with building dlang-bot. Rebasing or force pushing should fix the sqlite problem.

File: src/dmd/statementsem.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/statementsem.d:4493]: Please resist the urge to do anything more than move the code in this PR. Fixing other things is for separate PRs.

**Review by MoonlightSentinel — APPROVED** [review]

**Comment by benjones** [comment]
`/home/braddr/sandbox/at-client/pull-4888404-Linux_32_64/dmd/src/tests/cxxfrontend.c:573:1: fatal error: error writing to /tmp/ccPdYiHA.s: No space left on device
 }`

**Pull request merged**